### PR TITLE
new: support driverkit multiarch driver builds

### PR DESCRIPTION
**Pull request by FedeDP**
This PR adds support for driverkit configs $arch subfolders for newer driverversions, that we assume will be semversioned.  

On top of that, it also adds support for driverkit 0.9.1 features like `kernelurls` and multiarch builds.

Biggest challenge is keeping backward compatibility with current driverkit directory tree.

* Old directory tree:
`config/$driverversion/*.yml`
and expected url:
`https://download.falco.org/?prefix=driver/$driverversion/*.{o,ko}`

* New one:
`config/$driverversion(semver)/$arch/*.yml`
and expected url:
`https://download.falco.org/?prefix=driver/$driverversion/$arch/*.{o,ko}`

### CODE REVIEW

File: driverkit/Makefile [hunk 1]
```diff
@@ -5,33 +5,53 @@ ifeq ($(DRIVERKIT),)
 DRIVERKIT := "/bin/driverkit"
 endif
 
-CONFIGS := $(wildcard config/*/*.yaml)
+# Recursive wildcard
+rwildcard=$(foreach d,$(wildcard $(1:=/*)),$(call rwildcard,$d,$2) $(filter $(subst *,%,$2),$d))
+
+CONFIGS := $(call rwildcard,config/*,*.yaml)
```
Comment (FedeDP) [inline driverkit/Makefile:11]: Find all configs under config and its subfolders (eventually, for supported $arch).

File: driverkit/Makefile [hunk 2]
```diff
@@ -5,33 +5,53 @@ ifeq ($(DRIVERKIT),)
...
 TARGET_DISTRO ?= *
 TARGET_KERNEL ?= *
+TARGET_ARCH ?= *
+TARGET_HEADERS ?=
 S3_DRIVERS_BUCKET ?= "falco-distribution"
 S3_DRIVERS_KEY_PREFIX ?= "driver"
 SKIP_EXISTING ?= true
 
-validate: $(patsubst config_%,validate/%,$(subst /,_,$(wildcard config/${TARGET_VERSION}*/${TARGET_DISTRO}_${TARGET_KERNEL}-*)))
+validate: validate_old validate_new
```
Comment (FedeDP) [inline driverkit/Makefile:23]: Split in 2 sub targets to correctly manage new and old driverkit config tree.

File: driverkit/Makefile [hunk 3]
```diff
@@ -5,33 +5,53 @@ ifeq ($(DRIVERKIT),)
...
+# Not a big deal because we were not going to filter for TARGET_ARCH in any case.
+# TODO: rm once old format support is dropped
+validate_old: $(patsubst config_%,validate/%,$(subst /,_,$(wildcard config/${TARGET_VERSION}*/${TARGET_DISTRO}_${TARGET_KERNEL}-*)))
+# new fmt has $arch subfolders
+validate_new: $(patsubst config_%,validate/%,$(subst /,_,$(wildcard config/${TARGET_VERSION}*/${TARGET_ARCH}/${TARGET_DISTRO}_${TARGET_KERNEL}-*)))
 
 all: $(patsubst config_%,%,$(subst /,_,$(CONFIGS)))
 
-specific_target: $(patsubst config_%,%,$(subst /,_,$(wildcard config/${TARGET_VERSION}*/${TARGET_DISTRO}_${TARGET_KERNEL}-*)))
+specific_target: specific_target_old specific_target_new
```
Comment (FedeDP) [inline driverkit/Makefile:34]: Split in 2 sub targets to correctly manage new and old driverkit config tree.

File: driverkit/Makefile [hunk 4]
```diff
@@ -5,33 +5,53 @@ ifeq ($(DRIVERKIT),)
...
+specific_target_new: $(patsubst config_%,%,$(subst /,_,$(wildcard config/${TARGET_VERSION}*/${TARGET_ARCH}/${TARGET_DISTRO}_${TARGET_KERNEL}-*)))
 
 prepare: $(addprefix prepare_,$(VERSIONS))
 publish: $(addprefix publish_,$(VERSIONS))
 publish_s3: $(addprefix publish_s3_,$(VERSIONS))
 
 generate:
 	$(foreach VERSION,$(VERSIONS),\
-		utils/generate -k ${TARGET_KERNEL} -d ${TARGET_DISTRO} -v ${VERSION}; \
+		utils/generate -a $(TARGET_ARCH) -k ${TARGET_KERNEL} -d ${TARGET_DISTRO} -h ${TARGET_HEADERS} -v ${VERSION}; \
```
Comment (FedeDP) [inline driverkit/Makefile:49]: Generate target will now generate configs with proper `architecture` and `kernelurls` set.

File: driverkit/Makefile [hunk 5]
```diff
@@ -40,36 +60,53 @@ cleanup_s3:
 	S3_DRIVERS_BUCKET=${S3_DRIVERS_BUCKET} S3_DRIVERS_KEY_PREFIX=${S3_DRIVERS_KEY_PREFIX} utils/cleanup_s3 $(addprefix -v ,$(VERSIONS))
 
 # $(1): pseudo-target name
-# $(2): driver version
-# $(3): config file path
+# $(2): config file path
```
Comment (FedeDP) [inline driverkit/Makefile:63]: Dropped unused driver version param.

File: driverkit/Makefile [hunk 6]
```diff
@@ -40,36 +60,53 @@ cleanup_s3:
...
 # $(1): driver version
 define gen_publish_targets
 split_$(1)_kernelmodules:
+# root is old x86_64
+# TODO: rm once old format support is dropped
 ifneq ("$(wildcard output/$(1)/*.ko)","")
 	@mkdir -p output/$(1)/kernel-module
 	@mv -f output/$(1)/*.ko output/$(1)/kernel-module
 endif
+ifneq ("$(wildcard output/$(1)/x86_64/*.ko)","")
```
Comment (FedeDP) [inline driverkit/Makefile:86]: Basically, we now search in both root folder (old directory tree) and new $arch subfolders.

File: driverkit/utils/checkfiles [hunk 7]
```diff
@@ -8,6 +8,20 @@ source "${currentdir}/parseyaml"
...
+    IFS=/ read -a input_parts <<< ${1}
+    if [[ ${#input_parts[@]} -eq 4 ]]; then
+        arch="${input_parts[2]}/"
+    else
+        arch=""
+    fi
+    echo -n "${arch}"
+}
+
+arch="$(get_arch_from_path ${input})"
```
Comment (FedeDP) [inline driverkit/utils/checkfiles:23]: Fetch arch from config file path (if any); it is later used to check that if config file is under a $arch subfolder, its output module and probe must be under same $arch subfolder.

File: driverkit/utils/driverstats [hunk 8]
```diff
@@ -10,8 +10,14 @@ source "${currentdir}/parseyaml"
 TARGET_VERSION=${TARGET_VERSION:-"*"}
 TARGET_DISTRO=${TARGET_DISTRO:-"*"}
 TARGET_KERNEL=${TARGET_KERNEL:-"*"}
+TARGET_ARCH=${TARGET_ARCH:-"*"}
 
-configs=(config/${TARGET_VERSION}*/${TARGET_DISTRO}_${TARGET_KERNEL}-*.yaml)
+# Only load old configs if x86_64 or * arch is requested
+# TODO: rm this once old format support is dropped
+if [ "${TARGET_ARCH}" = "*" ] || [ "${TARGET_ARCH}" = "x86_64" ]; then
```
Comment (FedeDP) [inline driverkit/utils/driverstats:17]: If users filter for TARGET_ARCH, only load old config if either "*" (any) or "x86_64" is requested.

File: driverkit/utils/generate [hunk 9]
```diff
@@ -48,8 +59,39 @@ ensure_kernelrelease() {
...
+	esac
+}
+
+# If driver version is a semver, we are in the newest case
+# thus we must support %arch/ subfolder for configs.
+# Else, we are in the old case, and we only support x86_64 configs
+# in the TARGET_VERSION folder.
+# TODO: rm this once old format support is dropped, and always use ${TARGET_VERSION}/${TARGET_ARCH}
+function path_from_version() {
+  rx='^([0-9]+\.){0,2}(\*|[0-9]+)$'
```
Comment (FedeDP) [inline driverkit/utils/generate:83]: We assume that new driverversions, that will use new directory tree, all have a semver.

**Comment by FedeDP** [comment]
/cc @maxgio92 @zuc @leogr 

I think i will need multiple eyes on this one :)
Biggest problem is maintaining backward compatibility with current directory tree.

**Comment by leogr** [comment]
> /cc @maxgio92 @zuc @leogr
> 
> I think i will need multiple eyes on this one :) Biggest problem is maintaining backward compatibility with current directory tree.

Hey @FedeDP 

Could you describe the expected directory (and URL) layout in the PR description for a future reference? :pray: 
It would be also helpful to make sure we are on the same page.

**Comment by FedeDP** [comment]
@leogr i will edit the OP!

File: driverkit/Makefile (hunk 2 above)
Comment (maxgio92) [reply driverkit/Makefile:23]: I like it, even not elegant. When needed to cleanup it will be simple.

File: driverkit/utils/generate [hunk 10]
```diff
@@ -48,8 +59,39 @@ ensure_kernelrelease() {
     TARGET_KERNEL_RELEASE="${TARGET_KERNEL%_${TARGET_KERNEL_VERSION}}"
 }
 
+function arch_to_driverkit_arch() {
+	case "$1" in
+		"x86_64")
+			echo -n "amd64"
+			;;
+		"aarch64")
+			echo -n "arm64"
```
Comment (maxgio92) [inline driverkit/utils/generate:68]: Are we sure @FedeDP Driverkit always looks for "arm64", even though kernel headers packages could refer to "aarch64"?

Comment (FedeDP) [reply driverkit/utils/generate:68]: Yep because that's the golang naming :)

File: driverkit/Makefile (hunk 4 above)
Comment (maxgio92) [reply driverkit/Makefile:49]: See https://github.com/FedeDP/test-infra/pull/1 and review

File: driverkit/utils/generate [hunk 11]
```diff
@@ -8,8 +8,11 @@ CURRENT_DIR="$(pwd)"
 
 DEFAULT_KERNEL_VERSION="1"
 
-while getopts ":k:d:v:" arg; do
+while getopts ":a:k:d:v:h:" arg; do
   case $arg in
+    a)
+      TARGET_ARCH=${OPTARG}
```
[the remark this answers is not in the document]
Comment (FedeDP) [reply driverkit/utils/generate:14]: Uh yep i know that! But i don't think that we will ever call this without TARGET_ARCH set (i mean, with default "*" value); i assumed that generate target is just called by `scrape_and_generate`. Moreover, in README, i explicitly wrote
> Notice all the filters are optional (except for `generate` where they are mandatory).

I mean, if we can easily fix it, then big :+1: :)

[the remark this answers is not in the document]
Comment (maxgio92) [reply driverkit/utils/generate:14]: https://github.com/FedeDP/test-infra/pull/1 should sanitize "*" to actual supported architectures. BTW, it's a corner case

[the remark this answers is not in the document]
Comment (FedeDP) [reply driverkit/utils/generate:14]: Rebased on top of that!

**Comment by maxgio92** [comment]
I think we'll then need a refactoring of the DBG considering arbitrary architecture, maybe after the removal of deprecated (configuration and distribution) paths.

**Comment by FedeDP** [comment]
> I think we'll then need a refactoring of the DBG considering arbitrary architecture, maybe after the removal of deprecated (configuration and distribution) paths.

Yep; i think the work is in place indeed, we just need some improvements here and there (that will be much easier once we drop support for old tree paths)!

**Comment by FedeDP** [comment]
/hold
I'd wait until driverkit 0.9.2/0.10 is out, as it will bring fixes for qemu driver cross-compilation.

**Comment by FedeDP** [comment]
/unhold
Perhaps it is better to merge this and do a test build on a alpha driver git tag.
@leogr @maxgio92

**Comment by maxgio92** [comment]
/cc @leogr @zuc
LGTM let's coordinate first :-)

**Review by leogr — APPROVED** [review]
Let's give it a try :)
